<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Convert the coordinates of set of genes into vectors. — create_genesets • jazzPanda</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Convert the coordinates of set of genes into vectors. — create_genesets"><meta name="description" content="Convert the coordinates of set of genes into vectors."><meta property="og:description" content="Convert the coordinates of set of genes into vectors."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">jazzPanda</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/jazzPanda.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/phipsonlab/jazzPanda/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Convert the coordinates of set of genes into vectors.</h1>
      <small class="dont-index">Source: <a href="https://github.com/phipsonlab/jazzPanda/blob/main/R/create_genesets.R" class="external-link"><code>R/create_genesets.R</code></a></small>
      <div class="d-none name"><code>create_genesets.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convert the coordinates of set of genes into vectors.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_genesets</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">name_lst</span>,</span>
<span>  <span class="va">cluster_info</span>,</span>
<span>  <span class="va">sample_names</span>,</span>
<span>  <span class="va">bin_type</span>,</span>
<span>  <span class="va">bin_param</span>,</span>
<span>  <span class="va">w_x</span>,</span>
<span>  <span class="va">w_y</span>,</span>
<span>  use_cm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a SingleCellExperiment or SpatialExperiment or
SpatialFeatureExperiment object</p></dd>


<dt id="arg-name-lst">name_lst<a class="anchor" aria-label="anchor" href="#arg-name-lst"></a></dt>
<dd><p>A named list of strings giving the name of features that are
treated as background.</p></dd>


<dt id="arg-cluster-info">cluster_info<a class="anchor" aria-label="anchor" href="#arg-cluster-info"></a></dt>
<dd><p>A dataframe/matrix containing the centroid coordinates,
cluster and sample label for each cell.The column names must include
"x" (x coordinate), "y" (y coordinate),
"cluster" (cluster label) and "sample" (sample).</p></dd>


<dt id="arg-sample-names">sample_names<a class="anchor" aria-label="anchor" href="#arg-sample-names"></a></dt>
<dd><p>a vector of strings giving the sample names</p></dd>


<dt id="arg-bin-type">bin_type<a class="anchor" aria-label="anchor" href="#arg-bin-type"></a></dt>
<dd><p>A string indicating which bin shape is to be used for
vectorization. One of "square" (default), "rectangle", or "hexagon".</p></dd>


<dt id="arg-bin-param">bin_param<a class="anchor" aria-label="anchor" href="#arg-bin-param"></a></dt>
<dd><p>A numeric vector indicating the size of the bin. If the
<code>bin_type</code> is "square" or "rectangle", this will be a vector of length
two giving the numbers of rectangular quadrats in the x and y directions. If
the <code>bin_type</code> is "hexagonal", this will be a number giving the side
length of hexagons. Positive numbers only.</p></dd>


<dt id="arg-w-x">w_x<a class="anchor" aria-label="anchor" href="#arg-w-x"></a></dt>
<dd><p>A numeric vector of length two specifying the x coordinate
limits of enclosing box.</p></dd>


<dt id="arg-w-y">w_y<a class="anchor" aria-label="anchor" href="#arg-w-y"></a></dt>
<dd><p>A numeric vector of length two specifying the y coordinate
limits of enclosing box.</p></dd>


<dt id="arg-use-cm">use_cm<a class="anchor" aria-label="anchor" href="#arg-use-cm"></a></dt>
<dd><p>A boolean value that specifies whether to create spatial
vectors for genes using the count matrix and cell coordinates instead of
the transcript coordinates when both types of information are available.
The default setting is FALSE.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>A positive number specifying number of cores used for
parallelizing permutation testing. Default is one core
(sequential processing).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list of two matrices with the following components</p>
<dl><dt><code>gene_mt</code>  </dt>
<dd><p>contains the transcript count in each grid.
Each row refers to a grid, and each column refers to a gene.</p></dd>

<dt><code>cluster_mt</code>  </dt>
<dd><p>contains the number of cells in a specific
cluster in each grid. Each row refers to a grid, and each column refers
to a cluster.</p></dd>

</dl><p>The row order of <code>gene_mt</code> matches the row order of <code>cluster_mt</code>.</p>
<p>A matrix contains the sum count in each grid.
Each row refers to a grid, each column refers to a set in <code>name_lst</code>.
The column name will match the names in <code>name_lst</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SingleCellExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:SpatialFeatureExperiment’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     saveRDS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, saveRDS, setdiff, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">trans</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, min<span class="op">=</span><span class="fl">1</span>, max<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, min<span class="op">=</span><span class="fl">1</span>, max<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                feature_name<span class="op">=</span><span class="st">"A"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5</span>, min<span class="op">=</span><span class="fl">10</span>, max<span class="op">=</span><span class="fl">24</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">5</span>, min<span class="op">=</span><span class="fl">1</span>, max<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               feature_name<span class="op">=</span><span class="st">"B"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                         <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, min<span class="op">=</span><span class="fl">10</span>, max<span class="op">=</span><span class="fl">24</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10</span>, min<span class="op">=</span><span class="fl">10</span>, max<span class="op">=</span><span class="fl">24</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               feature_name<span class="op">=</span><span class="st">"C"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">trans</span><span class="op">$</span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">trans</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">trans</span><span class="op">$</span><span class="va">y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">trans</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">trans</span><span class="op">$</span><span class="va">cell</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell1"</span>,<span class="st">"cell2"</span>,<span class="st">"cell2"</span><span class="op">)</span>,replace<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                        size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">trans</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># create SpatialExperiment object</span></span></span>
<span class="r-in"><span><span class="va">trans_mol</span> <span class="op">&lt;-</span> <span class="fu">BumpyMatrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BumpyMatrix/man/splitAsBumpyMatrix.html" class="external-link">splitAsBumpyMatrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">trans</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span>, </span></span>
<span class="r-in"><span>    row <span class="op">=</span> <span class="va">trans</span><span class="op">$</span><span class="va">feature_name</span>, col <span class="op">=</span> <span class="va">trans</span><span class="op">$</span><span class="va">cell</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rep1_spe</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>     assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>molecules <span class="op">=</span> <span class="va">trans_mol</span><span class="op">)</span>,sample_id <span class="op">=</span><span class="st">"sample1"</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">geneset_res</span> <span class="op">&lt;-</span> <span class="fu">create_genesets</span><span class="op">(</span>x<span class="op">=</span><span class="va">rep1_spe</span>, sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                             name_lst<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dummy_A<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"C"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                             dummy_B<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                             bin_type<span class="op">=</span><span class="st">"square"</span>,</span></span>
<span class="r-in"><span>                             bin_param<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                             w_x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">25</span><span class="op">)</span>, w_y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">25</span><span class="op">)</span>, cluster_info<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Melody Jin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

